subsys/zigbee/cli/zigbee_cli_ctx_mngr.h

```diff
@@ -0,0 +1,98 @@
+#define ZIGBEE_CLI_CTX_MNGR_H__
+
+#include <shell/shell.h>
+
+#include "zigbee_cli_zdo_types.h"
+
+#define CTX_MNGR_ENTRY_IVALID_INDEX 0xFF
+
+
+/* Type of a frame a contex entry is associated with. */
+enum ctx_entry_type {
+	ZB_INVALID_ENTRY_TYPE,
+	ZB_ZDO_ENTRY_TYPE,
+	ZB_ZCL_ENTRY_TYPE,
+	ZB_PING_ENTRY_TYPE,
+};
+
+/* A context entry structure associated with a given frame, used to store
+ * a frame related data and a shell to be used for logging a data.
+ * Frame is identified by the type and its unique identifier number:
+ *     - ZDO frames are using the transaction number (TSN) as identifier number.
+ *     - ZCL frames are identified using the ZCL Sequence number.
+ *     - PING frames are using the custom counter, implemented along
+ *       the PING commands.
+ */
+struct ctx_entry {
+	bool taken;
+	/* Unique identifier of a frame. Type of used identifier depends on
+	 * the type of frame the entry is associated with.
+	 */
+	uint8_t frame_id;
+	enum ctx_entry_type entry_type;
+	/* Pointer to the shell to be used for printing logs. */
+	const struct shell *shell;
+	/* Union of Structures used to store a frame related data,
+	 * shared between the different frame types.
+	 */
+	union {
+		struct zdo_data zdo_data;
+	};
+};
+
+/**@brief Invalidate context manager entry to mark it free and prepare to be
+ *        reused later.
+ *
+ * @param ctx_entry       Pointer to the context manager entry to be freed.
+ */
+void ctx_mngr_invalidate_entry(struct ctx_entry *ctx_entry);
+
+/**@brief Get a pointer to a free context manager entry.
+ *
+ * @param entry_type      Type of a data to be stored in an entry.
+ *
+ * @return  Pointer to the context manager entry if a free entry was found,
+ *          NULL otherwise.
+ */
+struct ctx_entry *ctx_mngr_get_free_entry(enum ctx_entry_type entry_type);
+
+
+/**@brief Return a pointer to the context entry which is associated with
+ *        the given frame type and the given identifier.
+ *
+ * @param frame_id        Unique identifier of the frame the context entry
+ *                        is associated with. Type of the identifier used
+ *                        in the entry depends on the entry type.
+ *
+ * @param entry_type      Type of the frame the entry is associated with.
+ *
+ * @return  Pointer to the context manager entry if the entry was found,
+ *          NULL otherwise.
+ */
+struct ctx_entry *ctx_mngr_get_ctx_entry(uint8_t frame_id,
+					 enum ctx_entry_type entry_type);
+
+/**@brief Return an index of the given context entry.
+ *
+ * @param ctx_entry       Pointer to the context entry.
+ *
+ * @return  Index of given context entry.
+ */
+uint8_t ctx_mngr_get_index_of_ctx_entry(struct ctx_entry *ctx_entry);
+
+/**@brief Return a pointer to the context entry by given index.
+ *
+ * @param index           Index of the context entry.
+ *
+ * @return  Pointer to the context entry, NULL if not found.
+ */
+struct ctx_entry *ctx_mngr_get_ctx_entry_by_index(uint8_t index);
```

Comment:
What about hte second suggestion - do you have any comments?
I also have another proposal :slightly_smiling_face: - what do you think about using `ctx_mgr_find_entry` instead of `ctx_mngr_get_ctx_entry` ?
The reason why I am a bit skeptical about the current name is that the `get` does not imply if this is a new or existing entry.

Reply:
That's a good point